model/riscv_types.sail

```diff
@@ -367,26 +367,41 @@ mapping bool_not_bits : bool <-> bits(1) = {
   false  <-> 0b1
 }
 
-mapping size_bits : word_width <-> bits(2) = {
-  BYTE   <-> 0b00,
-  HALF   <-> 0b01,
-  WORD   <-> 0b10,
-  DOUBLE <-> 0b11
+val size_bits : word_width <-> bits(2)
+
+function size_bits_forwards width = match width {
+  BYTE   => 0b00,
+  HALF   => 0b01,
+  WORD   => 0b10,
+  DOUBLE => 0b11
+}
+
+function size_bits_backwards size = match size {
+  0b00 => BYTE,
+  0b01 => HALF,
+  0b10 => WORD,
+  0b11 => DOUBLE
```

Comment:
The version from PR #187 splits up this mapping into a forward and reverse function and adds some extra functions to handle quads. I can't say for sure why this was done, but I assume it is because word_width is currently 2 bits so cannot encode quads.

I'd be inclined to increase it to 3 bits and keep the handling of quads consistent with everything else. Most places size_bits is used it is prepended with a leading 0, so increasing it to 3 bits and giving quads an encoding of 100 should not impact these. For the couple of cases where it is not prepended with the leading 0 we can just pick off the bottom two bits.

I'd love any opinions on one approach vs the other.

Reply:
I think we should should increase to 3 bits for consistency.

Reply:
Thank you for pointing out! I am sorry that I forgot to leave a comment here. 

Yes, this was because wordwidth was 2 bits, and a few other components refers to word_width. I wasn't too sure about the semantics of other modules relying on word width. And the quad module doesn't rely on this function, as far as what I can recall. Therefore, I just did some minimal change to make my code compile, without changing the semantics.

Reply:
@martinberger I attempted to change this to 3 bits and am running into an issue with one particular encoding. All of the others had a 0 bit before so I just removed that. This encoding uses a sign bit, so we need to pull off the bottom two bits. I'm not sure how to do that inside of the encoding statement as I get an error when I try right now. Not sure if there is a different syntax that should be used here.

We could also look at splitting this (and probably the other mappings related to word_width) into two separate mappings—one for integer related instructions and one for floating point—if that would be preferable.

Code in question:
```
mapping clause encdec = LOAD(imm, rs1, rd, is_unsigned, size, false, false) if (word_width_bytes(size) < sizeof(xlen_bytes)) | (not(is_unsigned) & word_width_bytes(size) <= sizeof(xlen_bytes))
<-> imm @ rs1 @ bool_bits(is_unsigned) @ size_bits(size)[1..0] @ rd @ 0b0000011 if (word_width_bytes(size) < sizeof(xlen_bytes)) | (not(is_unsigned) & word_width_bytes(size) <= sizeof(xlen_bytes))
```

Error message:

```
 model/riscv_insts_base.sail:301.58-58:
301 |  <-> imm @ rs1 @ bool_bits(is_unsigned) @ size_bits(size)[1..0] @ rd @ 0b0000011 if (word_width_bytes(size) < sizeof(xlen_bytes)) | (not(is_unsigned) & word_width_bytes(size) <= sizeof(xlen_bytes))
    |                                                          ^
    | current token: [
make: *** [Makefile:261: generated_definitions/c/riscv_model_RV32.c] Error 1
```

Reply:
@jordancarlin OK, I see the problem. I think that is because [1..0] is not a bidirectional mapping. So if you go from right to left how would you get the third bit? You can do something like this
```
mapping strip : bits(2) <-> bits(3) = {
  0b00 <-> 0b000,
  0b01 <-> 0b001,
  0b10 <-> 0b010,
  0b11 <-> 0b011,
}
```
and then replace the `size_bits(size)[1..0]` by `strip(size_bits(size))` I think.

Reply:
@martinberger Thanks for the explanation. That builds and passes the test suite. 

That being said, this is making me wonder if having QUAD in size_enc makes sense in the first place. size_enc is currently only used by integer load/store and atomic instructions. The floating point extensions all have their widths hardcoded into the instructions (which seems like something that should be changed). Seeing as QUAD is only supported for floating point (and has the same encoding as unsigned byte for integer loads) it might make more sense to create a separate floating point size mapping down the road and leave this mapping just for integers. I know @Timmmm has been working on some related things if you want to weigh in on this at all.

Reply:
@jordancarlin I agreed that the question should `Quad be encoded needs careful though for the reasons you mention.

Reply:
Per RISC-V Golden Model Meeting, Quad will be added to word_width and size_enc will be extended to 3 bits to support it. This 3rd bit can be stripped away when needed (integer loads). This makes sense because quads are being added to amo and byte length floating point formats are expected.

Reply:
After the change, we should do a `grep` through the code where we use `word_width`, in particular where we make a pattern match against something of type `word_width` and add suitable cases either fall-though or matching against `QUAD` so the code is better documented. For example adding a suitable `internal_error(__FILE__, __LINE__, "...")`.

Reply:
Done. Thanks for mentioning that.